File: files/400-499/421/CH-421.py
import pandas as pd

path = "400-499/421/CH-421 Column Splitting.xlsx"
input = pd.read_excel(path, usecols="B", nrows=7, skiprows=2)
test = pd.read_excel(path, usecols="F:J", nrows=7, skiprows=2)


def _char_counts(x):
	if pd.isna(x):
		return []
	s = str(x)
	chars = list(dict.fromkeys(s))
	return [f"{ch}:{s.count(ch)}" for ch in chars]


id_col = "ID" if "ID" in input.columns else input.columns[0]
result = input[id_col].map(_char_counts).apply(pd.Series)
result = result.rename(columns=lambda c: f"ID{c + 1}")

print(result.equals(test))
# True

File: files/400-499/421/CH-421.R
library(tidyverse)
library(readxl)

path <- "400-499/421/CH-421 Column Splitting.xlsx"
input <- read_excel(path, range = "B3:B9")
test <- read_excel(path, range = "F3:J9")

result = input %>%
  transmute(
    out = map(ID, \(x) {
      chars <- unique(str_split_1(x, ""))
      str_c(chars, ":", str_count(x, fixed(chars)))
    })
  ) %>%
  unnest_wider(out, names_sep = "") %>%
  rename_with(\(x) str_replace(x, "^out", "ID"))

all.equal(result, test)
# [1] TRUE
